test/reduce.js

```diff
@@ -0,0 +1,392 @@
+
+module.exports = function reduce_test(testcase, minify_options, reduce_options) {
+    minify_options = minify_options || { compress: {}, mangle: false };
+    reduce_options = reduce_options || {};
+    var timeout = 1500; // start with a low timeout
+    var max_iterations = reduce_options.max_iterations || 1000;
+    var verbose = reduce_options.verbose;
+    var minify_options_json = JSON.stringify(minify_options);
+    var reduced = false;
+
+    if (testcase instanceof U.AST_Node) testcase = testcase.print_to_string();
+
+    // the initial timeout to assess the viability of the test case must be large
+    if (producesDifferentResultWhenMinified(testcase, minify_options, 5000)) {
+        // Replace expressions with constants that will be parsed into
+        // AST_Nodes as required.  Each AST_Node has its own permutation count,
+        // so these replacements can't be shared.
+        // Although simpler replacements are generally faster and better,
+        // feel free to experiment with a different replacement set.
+        var REPLACEMENTS = [
+            // "null", "''", "false", "'foo'", "undefined", "9",
+            "1", "0",
+        ];
+
+        // There's a relationship between each node's _permute counter and
+        // REPLACEMENTS.length which is why fractional _permutes were needed.
+        // One could scale all _permute operations by a factor of `steps`
+        // to only deal with integer operations, but this works well enough.
+        var steps = 4;        // must be a power of 2
+        var step = 1 / steps; // 0.25 is exactly representable in floating point
+
+        var tt = new U.TreeTransformer(function(node, descend, in_list) {
+            if (CHANGED) return;
+
+            // quick ignores
+            if (node instanceof U.AST_Toplevel) return;
+            if (node instanceof U.AST_Accessor) return;
+            if (node instanceof U.AST_Directive) return;
+            // if (node instanceof U.AST_Var) return;
+            // if (node instanceof U.AST_VarDef) return;
+
+            var parent = tt.parent();
+
+            // ignore call expressions
+            if (parent instanceof U.AST_Call && parent.expression === node) return;
+
+            // ensure that the _permute prop is a number.
+            // can not use `node.start._permute |= 0;` as it will erase fractional part.
+            if (typeof node.start._permute === "undefined") node.start._permute = 0;
+
+            // if node reached permutation limit - skip over it.
+            // no structural AST changes before this point.
+            if (node.start._permute >= REPLACEMENTS.length) return;
+
+            if (parent instanceof U.AST_Assign
+                    && parent.left === node
+                || parent instanceof U.AST_Unary
+                    && parent.expression === node
+                    && ["++", "--", "delete"].indexOf(parent.operator) >= 0) {
+                // ignore lvalues
+                return;
```

Comment:
Should this block of code should go before that `// no structural AST changes before this point.` comment? 🤔

Reply:
I suppose, but it's not significant - the cost of an `if`. A single AST iteration sandbox run is literally a million times more expensive.

Reply:
Keep in mind that only the two of us will ever use it. Judge this utility by ufuzz code quality, not Compressor code quality. 
;-)

Reply:
You may be tempted to introduce AST optimizations into this code - resist that urge. The entire point is to make stupid - and even unreasonable - changes and see if it works. Introducing AST optimizations may actually be counterproductive here. You will notice in a few places that I only interrogate and return the first item in an array of elements. The reason is because other code at the bottom of the transformer speculatively gets rid of uneeded list elements.